smarts/core/renderer.py

```diff
@@ -0,0 +1,263 @@
+        self.teardown()
+        self._showbase_instance = None
+
+    def __del__(self):
+        self.destroy()
+
+    def create_vehicle(self, glb_model: str, vid: str, color, pose: Pose):
+        with pkg_resources.path(models, glb_model) as path:
+            node_path = self._showbase_instance.loader.loadModel(str(path.absolute()))
+        node_path.setName("vehicle-%s" % vid)
+        node_path.setColor(color)
+        pos, heading = pose.as_panda3d()
+        node_path.setPosHpr(*pos, heading, 0, 0)
+        node_path.hide(RenderMasks.DRIVABLE_AREA_HIDE)
+        return node_path
+
+    def begin_rendering_vehicle(self, vehicle_path: NodePath, is_agent: bool):
+        """ adds the vehicle node to the scene graph """
+        vehicle_path.reparentTo(self._vehicles_np if is_agent else self._root_np)
+
+    class OffscreenCamera(NamedTuple):
+        camera_np: NodePath
+        buffer: GraphicsOutput
+        tex: Texture
+        render_engine: _ShowBaseInstance
+
+        def teardown(self):
+            self.camera_np.removeNode()
+            region = self.buffer.getDisplayRegion(0)
+            region.window.clearRenderTextures()
+            self.buffer.removeAllDisplayRegions()
+            self.render_engine.graphicsEngine.removeWindow(self.buffer)
+
+    def build_offscreen_camera(
+        self,
+        name: str,
+        mask: int,
+        width: int,
+        height: int,
+        resolution: float,
+    ):
+        # setup buffer
+        win_props = WindowProperties.size(width, height)
+        fb_props = FrameBufferProperties()
+        fb_props.setRgbColor(True)
+        fb_props.setRgbaBits(8, 8, 8, 1)
+        # XXX: Though we don't need the depth buffer returned, setting this to 0
+        #      causes undefined behaviour where the ordering of meshes is random.
+        fb_props.setDepthBits(8)
+
+        buffer = self._showbase_instance.win.engine.makeOutput(
+            self._showbase_instance.pipe,
+            "{}-buffer".format(name),
+            -100,
+            fb_props,
+            win_props,
+            GraphicsPipe.BFRefuseWindow,
+            self._showbase_instance.win.getGsg(),
+            self._showbase_instance.win,
+        )
+        buffer.setClearColor((0, 0, 0, 0))  # Set background color to black
+
+        # setup texture
+        tex = Texture()
+        region = buffer.getDisplayRegion(0)
+        region.window.addRenderTexture(
+            tex, GraphicsOutput.RTM_copy_ram, GraphicsOutput.RTP_color
+        )
+
+        # setup camera
+        lens = OrthographicLens()
+        lens.setFilmSize(width * resolution, height * resolution)
+
+        camera_np = self._showbase_instance.makeCamera(
+            buffer, camName=name, scene=self._root_np, lens=lens
+        )
+        camera_np.reparentTo(self._root_np)
+
+        # mask is set to make undesireable objects invisible to this camera
+        camera_np.node().setCameraMask(camera_np.node().getCameraMask() & mask)
```

Comment:
I think my concern here is if this camera will see all vehicles even if it is from other SMARTS instances. I think that is the case with the current configuration.

Reply:
I thought the `scene=self._root_np` part of the `makeCamera()` call would limit what the camera sees to just the vehicle's scene.  However, I'm not sure of this (the documentation is unclear) and couldn't test it.  If so, this is different than the current situation because there we only have one root node under `render` called "sim", whereas here, each simulation has its own uniquely-named root node.

Reply:
Regarding your concern, notwithstanding my comment which I think means the scene graphs can be kept separate for different SMARTS instances, I do think it's valid:  that is, I don't think we will ever be able to overlap multiple, simultaneous SMARTS instances until we (a) make our own simulation clock in SMARTS cf. the comment I added there in `step()` and (b) figure out a way to cause ShowBase to render only one branch of the scene graph at a time.  I just added a couple more comments here in the Renderer class to try to clarify that.

Reply:
I think there are a few things I can address here. Cameras in game engines are attached to a node in order to inherit the transformations (and other attributes) of the parent node. This generally means if the parent note is transformed (e.g. translation) the camera will maintain relative distance. The camera will still see all other renderable nodes regardless of existing on a different branch of the scene graph. From what I can see all of the nodes are connected via the the common root of the showbase `render` node.

I think there may be a few possible workarounds. But using `hide()` on each of the other simulation roots before rendering and then calling `show()` after might work best: https://docs.panda3d.org/1.10/python/reference/panda3d.core.NodePath?highlight=nodepath#panda3d.core.NodePath.hide

Reply:
I think `Camera.setScene` is the same as passing the `scene` parameter to the `makeCamera()` call, as I'm already doing. So I thnk we are already good because there is currently a difffernt scene root (all children of `render`) for each simulation.  But I can add the `hide()` and `show()` calls to `Renderer.render()`, or try with masks, if you like that alternative better.  However, note that all of these only solve issue (b) of my previous comment, issue (a) will still exist until we fix the clock.  If you think it's not biting off too much, I can go ahead and try to fix the clock too with this PR.